extensions/interactions/RatioExpressionInput/directives/ratio-expression-input-validation.service.ts

```diff
@@ -0,0 +1,216 @@
+    warningsList = warningsList.concat(
+      this.getCustomizationArgsWarnings(customizationArgs));
+
+    warningsList = warningsList.concat(
+      this.baseInteractionValidationServiceInstance.getAllOutcomeWarnings(
+        answerGroups, defaultOutcome, stateName));
+
+    let rulesService = this;
+
+    // Checks whether currentInput is in simplest form or not.
+    let isInSimplestForm = function(
+        currentRuleType: string,
+        ratio: Ratio,
+        currentInput: number[]
+    ): boolean {
+      return currentRuleType === 'IsEquivalent' && (
+        !rulesService.rof.arrayEquals(
+          ratio.convertToSimplestForm(), currentInput));
+    };
+
+    // Checks whether currentInput has less number of terms than seenInput.
+    let hasLessNumberOfTerms = function(
+        currentRuleType: string,
+        seenRuleType: string,
+        currentInput: number,
+        seenInput: number[]
+    ): boolean {
+      return currentRuleType === 'HasNumberOfTermsEqualTo' &&
+      seenRuleType !== 'HasNumberOfTermsEqualTo' && (
+        ratioRulesService.HasNumberOfTermsEqualTo(
+          seenInput, {y: currentInput}));
+    };
+
+    // The following validations ensure that there are no redundant rules
+    // present in the answer groups. In particular, an Equals rule will make
+    // all of the following rules with a matching input invalid. A
+    // HasNumberOfTermsEqualTo rule will make the following rules of the same
+    // rule type and a matching input invalid.
+    let seenRules = [];
+
+    for (let i = 0; i < answerGroups.length; i++) {
+      let rules = answerGroups[i].getRulesAsList();
+      for (let j = 0; j < rules.length; j++) {
+        let currentRuleType = <string> rules[j].type;
+        let currentInput = null;
+        var ratio: Ratio = null;
+        if (currentRuleType === 'HasNumberOfTermsEqualTo') {
+          currentInput = <number> rules[j].inputs.y;
+        } else {
+          currentInput = <number[]> rules[j].inputs.x;
+        }
+
+        if (minimumNumberOfTerms > 2) {
+          if (currentRuleType === 'HasNumberOfTermsEqualTo') {
+            if (currentInput < minimumNumberOfTerms) {
+              warningsList.push({
+                type: AppConstants.WARNING_TYPES.ERROR,
+                message: (
+                  `Rule ${j + 1} from answer group ${i + 1} will never be` +
+                ' matched because it has fewer number of terms than required.')
+              });
+            }
+          } else {
+            ratio = rulesService.rof.fromList(<number[]> currentInput);
+            if (ratio.getNumberOfTerms() < minimumNumberOfTerms) {
+              warningsList.push({
+                type: AppConstants.WARNING_TYPES.ERROR,
+                message: (
+                  `Rule ${j + 1} from answer group ${i + 1} will never be` +
+                ' matched because it has fewer number of terms than required.')
```

Comment:
This & above need to be column aligned like the strings below.

Reply:
done